pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,318 @@
+	var epData *endpointMetadata
+	var err error
+
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	if len(endpoint.Networking.Addressing) == 0 {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("Failed to get valid endpoint IPs, skipping update to egress policy.")
+		return
+	}
+
+	if epData, err = manager.getEndpointMetadata(endpoint); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("Failed to get valid endpoint metadata, skipping update to egress policy.")
+		return
+	}
+
+	// Remove old: check if the endpoint was previously selected by any of the policies.
+	if policies, ok := manager.policyEndpoints[epData.id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			err := manager.deleteEgressMap(config, epData)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.K8sEndpointName: endpoint.Name,
+					logfields.K8sNamespace:    endpoint.Namespace,
+				}).Error("Error updating endpoint mapping.")
+				return
+			}
+		}
+		delete(manager.policyEndpoints, epData.id)
+		delete(manager.epDataStore, epData.id)
+	}
+
+	// Upsert new: check if current policies select new endpoint. Also updates endpoint cache
+	manager.epDataStore[epData.id] = epData
+	for _, config := range manager.policyConfigs {
+		if config.policyConfigSelectsEndpoint(epData) {
+			err := manager.upsertPolicyEndpoint(config, epData)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.K8sEndpointName: endpoint.Name,
+					logfields.K8sNamespace:    endpoint.Namespace,
+				}).Error("Error upserting pod mapping for pod.")
+				return
+			}
+		}
+	}
+}
+
+// OnDeleteEndpoint is the event handler for endpoint deletions.
+func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	var epData *endpointMetadata
+	var err error
+	if epData, err = manager.getEndpointMetadata(endpoint); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("Failed to get valid endpoint metadata, abort deleting endpoint mapping.")
+		return
+	}
+
+	if policies, ok := manager.policyEndpoints[id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			manager.deleteEgressMap(config, epData)
+		}
+		delete(manager.policyEndpoints, id)
+	}
+	delete(manager.epDataStore, id)
+}
+
+func (manager *Manager) getEndpointMetadata(endpoint *k8sTypes.CiliumEndpoint) (*endpointMetadata, error) {
```

Comment:
This function doesn't need to have the manager as the receiver.

Reply:
fixed.